### PR TITLE
Add endpoint for push leaderboard scores.

### PR DESCRIPTION
**Pull request by Andrei-Dolgolev**
## How to test these changes?

Whitelisted endpoints:
```bash
### Get leaderboard paginated limit offset
GET <base_engine_url>/leaderboard/?leaderboard_id=<leaderboard_id>&limit=<10>&offset=<0>

### Return 3 quatiles 25, 50, 75
GET <base_engine_url>/leaderboard/quartiles/?leaderboard_id=<leaderboard_id>

### return count of addresses in leaderboard
GET <base_engine_url>/leaderboard/count/addresses/?leaderboard_id=<leaderboard_id>

### Return address data if that address in leaderboard and closes by rating addresses depending on window size if `window_size=0` return user data only
GET <base_engine_url>/leaderboard/position/?leaderboard_id=<leaderboard_id>&address=<user_address>&window_size=<1>&limit=<10>&offset=<0>
```

Add new scores to leaderboard:
```bash
PUT <base_engine_url>/leaderboard/{leaderboard_id}/scores?overwrite=<bool>
Authorization: Bearer <token>

[
        {
            "address": <address>,
            "score": 200000000,
            "points_data": {
                "some_points": 100000000,
                "some_points_type_2": 100000000
        }
   }
]
```

`overwrite=True` will delete all records from leaderboard before insert.
`overwrite=True` Will update `points_data` and score if that address already exists in leaderboard.

required enviroment variables:
```
MOONSTREAM_APPLICATION_ID
MOONSTREAM_ACCESS_ADMIN_TOKEN
```

## Added CLI for manage leaderboard resources: 

Add access resource if it not exists:
```bash
engineapi engine-db leaderboard permissions assign  --leaderboard-id <leaderboard_UUID>
```
Get resources for leaderboards:
```bash
engineapi engine-db leaderboard permissions list
```

Remove resource from leaderboard:
```bash
engineapi engine-db leaderboard permissions revoke --leaderboard-id <leaderboard_UUID>
```

## REQUIRED database migration:
```bash
./alembic.sh -c alembic.dev.ini upgrade head
```

### CODE REVIEW

**Comment by Andrei-Dolgolev** [comment]
Instead of using `BroodAuthMiddlelayer` we will create resource wich will resolve permission to write to table.

It's required from us add `resource_id` column for leaderboard table in engine database. So when user have write permisson he will be able push updatets.

File: api/setup.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
     packages=find_packages(),
     install_requires=[
         "boto3",
+        "bugout>=0.1.19",
```
Comment (kompotkot) [inline api/setup.py:16]: Use new version

**Review by kompotkot — CHANGES_REQUESTED** [review]

File: api/engineapi/actions.py [hunk 2]
```diff
@@ -1097,3 +1115,98 @@ def add_scores(db_session: Session, leaderboard_id, scores):
...
+
+    leaderboard = (
+        db_session.query(Leaderboard).filter(Leaderboard.id == leaderboard_id).one()
+    )
+
+    if leaderboard.resource_id is not None:
+
+        raise Exception("Leaderboard already has a resource")
+
+    if resource_id:
```
Comment (kompotkot) [inline api/engineapi/actions.py:1161]: If resource_id is not None

File: api/engineapi/actions.py [hunk 3]
```diff
@@ -1097,3 +1115,98 @@ def add_scores(db_session: Session, leaderboard_id, scores):
...
+        token=MOONSTREAM_ACCESS_ADMIN_TOKEN,
+        application_id=MOONSTREAM_APPLICATION_ID,
+        resource_data=resource_data,
+        timeout=10,
+    )
+    return resource
+
+
+def assign_resource(
+    db_session: Session, leaderboard_id: uuid.UUID, resource_id: Optional[uuid.UUID]
```
Comment (kompotkot) [inline api/engineapi/actions.py:1146]: `resource_id: Optional[uuid.UUID] = None`

File: api/engineapi/cli.py [hunk 4]
```diff
@@ -328,7 +329,8 @@ def add_scores_handler(args: argparse.Namespace) -> None:
                 scores = actions.add_scores(
                     db_session=db_session,
                     leaderboard_id=args.leaderboard_id,
-                    scores=json_input,
+                    scores=[data.Score(**score) for score in json_input],
```
Comment (kompotkot) [inline api/engineapi/cli.py:332]: Better to prepare variable and return readable in log message that `json.load(f)` not parsed in `data.Score()`

File: api/engineapi/cli.py [hunk 5]
```diff
@@ -380,6 +382,58 @@ def create_leaderboard_handler(args: argparse.Namespace) -> None:
...
+        print(resources)
+
+
+def revoke_resource_handler(args: argparse.Namespace) -> None:
+    with db.yield_db_session_ctx() as db_session:
+        try:
+            resource = actions.revoke_resource(
+                db_session=db_session,
+                leaderboard_id=args.leaderboard_id,
+            )
```
Comment (kompotkot) [inline api/engineapi/cli.py:417]: Print at least "complete"

File: api/engineapi/cli.py [hunk 6]
```diff
@@ -380,6 +382,58 @@ def create_leaderboard_handler(args: argparse.Namespace) -> None:
...
+            logger.error(f"Unhandled /assign_resource exception: {err}")
+            return
+
+
+def list_resources_handler(args: argparse.Namespace) -> None:
+
+    with db.yield_db_session_ctx() as db_session:
+        resources = actions.list_leaderboards_resources(db_session=db_session)
+
+        print(resources)
```
Comment (kompotkot) [inline api/engineapi/cli.py:408]: In function above you are logging, here printing, better to use one common approach.

File: api/engineapi/cli.py [hunk 7]
```diff
@@ -519,12 +573,108 @@ def main() -> None:
     parser_leaderboard_score.add_argument(
         "--input-file",
         type=str,
-        required=True,
+        required=False,
```
Comment (kompotkot) [inline api/engineapi/cli.py:576]: Just delete this line, not dequired by default

File: api/engineapi/settings.py [hunk 8]
```diff
@@ -139,3 +147,10 @@
     raise ValueError(
         f"ENGINE_DB_POOL_RECYCLE_SECONDS must be an integer: {ENGINE_DB_POOL_RECYCLE_SECONDS_RAW}"
     )
+
+MOONSTREAM_APPLICATION_ID = os.environ.get("MOONSTREAM_APPLICATION_ID", "")
+
+
+LEADERBOARD_RESOURCE_TYPE = "leaderboard"
+
+MOONSTREAM_ACCESS_ADMIN_TOKEN = os.environ.get("MOONSTREAM_ACCESS_ADMIN_TOKEN", "")
```
Comment (kompotkot) [inline api/engineapi/settings.py:156]: We do not have this variable, we have variable:
```
MOONSTREAM_ADMIN_ACCESS_TOKEN
```

Comment (kompotkot) [reply api/engineapi/settings.py:156]: Also add this in `sample.env`

**Comment by kompotkot** [comment]
@bugout-dev check
- [ ] Run migration

File: api/engineapi/middleware.py [hunk 9]
```diff
@@ -102,3 +102,38 @@ def __init__(
...
+        method = request.method
+        if path in self.whitelist.keys() and self.whitelist[path] == method:
+            return await call_next(request)
+
+        authorization_header = request.headers.get("authorization")
+        if authorization_header is None:
+            return Response(
+                status_code=403, content="No authorization header passed with request"
+            )
+        user_token_list = authorization_header.split()
```
Comment (kompotkot) [inline api/engineapi/middleware.py:132]: This is not user_token_list, this is components of auth header (bearer and token), better to call it 
`authorization_header_components`

File: api/engineapi/middleware.py [hunk 10]
```diff
@@ -102,3 +102,38 @@ def __init__(
...
+        if len(user_token_list) != 2:
+            return Response(status_code=403, content="Wrong authorization header")
+        user_token: str = user_token_list[-1]
+
+        request.state.token = user_token
```
Comment (kompotkot) [inline api/engineapi/middleware.py:137]: Why we store this in state? I checked code and we no where are using `user_token`.. If then, why do we need this middleware, we also not calling any authorization places

**Review by kompotkot — CHANGES_REQUESTED** [review]

File: api/setup.py (hunk 1 above)
Comment (Andrei-Dolgolev) [reply api/setup.py:16]: Add 0.2.2

File: api/engineapi/actions.py [hunk 11]
```diff
@@ -1097,3 +1115,98 @@ def add_scores(db_session: Session, leaderboard_id, scores):
     db_session.commit()
```
[the remark this answers is not in the document]
Comment (Andrei-Dolgolev) [reply api/engineapi/actions.py:1116]: Idea make changes in 1 commit for data be consistent

File: api/engineapi/actions.py [hunk 12]
```diff
@@ -1097,3 +1115,98 @@ def add_scores(db_session: Session, leaderboard_id, scores):
...
+    token: Optional[uuid.UUID] = None,
+) -> BugoutResource:
+
+    resource_data: Dict[str, Any] = {
+        "type": LEADERBOARD_RESOURCE_TYPE,
+        "leaderboard_id": leaderboard_id,
+    }
+
+    if token is None:
+        token = MOONSTREAM_ACCESS_ADMIN_TOKEN
```
[the remark this answers is not in the document]
Comment (Andrei-Dolgolev) [reply api/engineapi/actions.py:1135]: I actually was think create resource maybe just with admin and grant read access to use what you think.

File: api/engineapi/middleware.py (hunk 10 above)
Comment (Andrei-Dolgolev) [reply api/engineapi/middleware.py:137]: Added in last commit we start use that token in handler.

**Review by kompotkot — APPROVED** [review]
lg

**Pull request merged**